src/sources/file_descriptors/pipe.rs

```diff
@@ -0,0 +1,191 @@
+use std::fs::File;
+use std::io;
+use std::os::raw::c_int;
+use std::os::unix::io::FromRawFd;
+
+use super::{file_descriptor_source, FileDescriptorConfig};
+use codecs::decoding::{DeserializerConfig, FramingConfig};
+use indoc::indoc;
+use vector_config::configurable_component;
+use vector_core::config::LogNamespace;
+
+use crate::{
+    config::{GenerateConfig, Output, Resource, SourceConfig, SourceContext, SourceDescription},
+    serde::default_decoding,
+    shutdown::ShutdownSignal,
+    SourceSender,
+};
+/// Configuration for the `pipe` source.
+#[configurable_component(source)]
+#[derive(Clone, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PipeConfig {
+    /// The maximum buffer size, in bytes, of incoming messages.
+    ///
+    /// Messages larger than this are truncated.
+    #[serde(default = "crate::serde::default_max_length")]
+    pub max_length: usize,
+
+    /// Overrides the name of the log field used to add the current hostname to each event.
+    ///
+    /// The value will be the current hostname for wherever Vector is running.
+    ///
+    /// By default, the [global `host_key` option](https://vector.dev/docs/reference/configuration//global-options#log_schema.host_key) is used.
+    pub host_key: Option<String>,
+
+    #[configurable(derived)]
+    pub framing: Option<FramingConfig>,
+
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    pub decoding: DeserializerConfig,
+
+    /// The file descriptor number to read from.
+    pub fd: c_int,
```

Comment:
I'm not sure this is the right type. At minimum it should be unsigned, as negative file descriptors are invalid. I think it should be a standard type such as `u16` or `u32` which would be converted into `c_int` as needed below, but I'm open to disagreement on that. If we want to disallow using this source in place of the `stdin` source, we could further use `NonZeroU16` to forbid zero as well.

Reply:
Would you be fine with `c_uint`?  Would guarantee it can hold any value that the underlying API can hold via `c_int`.

Reply:
It's better, but I'm still not entirely behind that. They underlying `unsigned int` will never be narrower than 32-bits on any platform vector will ever support, and FD numbers are usually restricted to well under the 16-bit range. As such, `u32` or `NonZeroU32` are more appropriate. This avoids exposing platform-specific types in the configuration interface.